asmopt/tests/optimizer.rs

```diff
@@ -0,0 +1,311 @@
+        // END BATCH Unimplemented
+        return;
+        // END BATCH
+    }
+}
+"#;
+
+    let parsed = parse_asm(None, input).unwrap();
+    let analyzed = analyze(parsed).unwrap();
+    let optimized = optimize(analyzed).to_string();
+    assert_eq!(optimized, expectation);
+}
+
+#[test]
+fn keep_machine_with_multiple_references() {
+    let input = r#"
+    machine Main with degree: 8 {
+        Helper helper;
+        
+        reg pc[@pc];
+        reg X[<=];
+        reg Y[<=];
+        reg A;
+
+        // Two different instructions using the same machine
+        instr double X -> Y link => Y = helper.double(X);
+        instr triple X -> Y link => Y = helper.triple(X);
+
+        function main {
+            // Only using one instruction
+            A <== double(2);
+            return;
+        }
+    }
+
+    machine Helper with degree: 8 {
+        reg pc[@pc];
+        reg X[<=];
+        reg Y[<=];
+
+        function double x: field -> field { return x + x; }
+        function triple x: field -> field { return x + x + x; }
+    }
+    "#;
+
+    let expectation = r#"machine Main with degree: 8 {
+    ::Helper helper
+    reg pc[@pc];
+    reg X[<=];
+    reg Y[<=];
+    reg A;
+    instr double X -> Y link => Y = helper.double(X){  }
+    function main {
+        A <=Y= double(2);
+        // END BATCH Unimplemented
+        return;
+        // END BATCH
+    }
+}
+machine Helper with degree: 8 {
+    reg pc[@pc];
+    function double x: field -> field {
+        return x + x;
+        // END BATCH
+    }
+    function triple x: field -> field {
+        return x + x + x;
+        // END BATCH
+    }
+}
+"#;
+
+    let parsed = parse_asm(None, input).unwrap();
+    let analyzed = analyze(parsed).unwrap();
+    let optimized = optimize(analyzed).to_string();
+    assert_eq!(optimized, expectation);
+}
+
+#[test]
+fn keep_machine_parameters() {
+    let input = r#"
+    machine Main with degree: 8 {
+        Required required;
+        ParamMachine sub(required);
+        Unused unused;
+        
+        reg pc[@pc];
+        reg X[<=];
+        reg Y[<=];
+        reg A;
+
+        instr compute X -> Y link => Y = sub.compute(X);
+
+        function main {
+            A <== compute(1);
+            return;
+        }
+    }
+
+    machine ParamMachine(mem: Required) with degree: 8 {
+        reg pc[@pc];
+        reg X[<=];
+        reg Y[<=];
+
+        function compute x: field -> field {
+            return x + x;
+        }
+    }
+
+    machine Required with
+        latch: latch,
+        operation_id: operation_id
+    {
+        operation compute<0> x -> y;
+
+        col fixed latch = [1]*;
+        col witness operation_id;
+        col witness x;
+        col witness y;
+        
+        y = x + x;
+    }
+
+    machine Unused with degree: 8 {
+        reg pc[@pc];
+        col witness w;
+        w = w * w;
+    }
+    "#;
+
+    let expectation = r#"machine Main with degree: 8 {
+    ::Required required
```

Comment:
Shouldn't `Required` be removed here? It is not used anymore

Reply:
In fact, it's being used, it just doesn't print (I'm going to open an issue on this). That's partly why I created the constrained_machines that are no longer needed (because they really weren't).

Reply:
mmm it's required because its a parameter to the submachine?

Reply:
Oh, sorry, yes, exactly. Correctly printed, this should look like this:

```
::Required required
::ParamMachine sub(required)
```